### PR TITLE
REST APIs for Gobblin Service job configuration.

### PR DESCRIPTION
**Pull request by htran1**
Create APIs for create, get, full update, and delete of JobConfigs.
JobConfigs are turned into JobSpecs stored in a MutableJobCatalog.

### CODE REVIEW

**Comment by htran1** [comment]
@ibuenros can  you review?

File: gobblin-restli/gobblin-job-config-service/api/src/main/pegasus/gobblin/service/JobConfig.pdsc [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
```
Comment (ibuenros) [inline gobblin-restli/gobblin-job-config-service/api/src/main/pegasus/gobblin/service/JobConfig.pdsc]: Which fields are required and which are optional?

File: gobblin-restli/gobblin-job-config-service/api/src/main/pegasus/gobblin/service/JobConfig.pdsc [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    },
+    {
+      "name" : "jobGroup",
+      "type" : "string",
+      "doc" : "Group of the job"
+    },
+    {
+      "name" : "schedule",
+      "type" : "string",
+      "doc" : "Schedule for job"
```
Comment (ibuenros) [inline gobblin-restli/gobblin-job-config-service/api/src/main/pegasus/gobblin/service/JobConfig.pdsc]: Can you make it clear what the format of the schedule string is? (e.g. cron-style)

File: gobblin-restli/gobblin-job-config-service/api/src/main/pegasus/gobblin/service/JobConfig.pdsc [hunk 3]
```diff
@@ -0,0 +1,32 @@
+{
+  "type" : "record",
+  "name" : "JobConfig",
+  "namespace": "gobblin.service",
+  "doc" : "Configuration for a job",
+  "fields" : [
```
Comment (ibuenros) [inline gobblin-restli/gobblin-job-config-service/api/src/main/pegasus/gobblin/service/JobConfig.pdsc]: I think we also need a job URI. Ultimately, this is the only guaranteed unique identifier.

File: gobblin-restli/gobblin-job-config-service/api/src/main/pegasus/gobblin/service/JobConfigId.pdsc [hunk 4]
```diff
@@ -0,0 +1,17 @@
+{
+  "type" : "record",
+  "name" : "JobConfigId",
+  "namespace": "gobblin.service",
+  "doc" : "Identifier for a job",
+  "fields" : [
+    {
```
Comment (ibuenros) [inline gobblin-restli/gobblin-job-config-service/api/src/main/pegasus/gobblin/service/JobConfigId.pdsc]: The main return value here should be a job URI. The rest is just metadata.

File: gobblin-restli/gobblin-job-config-service/server/src/main/java/gobblin/service/JobConfigServer.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Server that handles Rest.li requests over HTTP to configure Gobblin Service jobs
+ */
+public class JobConfigServer {
```
Comment (ibuenros) [inline gobblin-restli/gobblin-job-config-service/server/src/main/java/gobblin/service/JobConfigServer.java]: There is a class being reviewed on another PR that sets up a generic `EmbeddedRestliServer`. I'd prefer if you wait for that to be ready and either make this an extension of that one, or not have this class at all.

File: gobblin-restli/gobblin-job-config-service/server/src/main/java/gobblin/service/JobConfigsResource.java [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+
+@RestLiCollection(name = "jobconfigs", namespace = "gobblin.service")
+/**
+ * Rest.li resource for handling job configuration requests
+ */
+public class JobConfigsResource extends ComplexKeyResourceTemplate<JobConfigId, EmptyRecord, JobConfig> {
+  private static final Logger LOG = LoggerFactory.getLogger(JobConfigsResource.class);
+
+  private MutableJobCatalog jobCatalog;
```
Comment (ibuenros) [inline gobblin-restli/gobblin-job-config-service/server/src/main/java/gobblin/service/JobConfigsResource.java]: You should add an `@Inject` annotation. That's the way rest.li resources are supposed to work.

File: gobblin-restli/gobblin-job-config-service/server/src/main/java/gobblin/service/JobConfigsResource.java [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+  JobConfigsResource(FSJobCatalog jobCatalog) {
```
Comment (ibuenros) [inline gobblin-restli/gobblin-job-config-service/server/src/main/java/gobblin/service/JobConfigsResource.java]: AFAIK rest.li requires a default constructor.

File: gobblin-restli/gobblin-job-config-service/server/src/main/java/gobblin/service/JobConfigsResource.java [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+    LOG.info("Get called with jobGroup " + jobGroup + " jobName " + jobName);
+
+    try {
+      JobSpec jobSpec = jobCatalog.getJobSpec(new URI("/" + jobGroup  + "/" + jobName));
+      Properties jobProps = jobSpec.getConfigAsProperties();
+      String schedule = jobProps.getProperty(ConfigurationKeys.JOB_SCHEDULE_KEY);
+      String template = jobProps.getProperty(ConfigurationKeys.JOB_TEMPLATE_PATH);
+      // remove keys that were injected by jobCatalog or as part of jobSpec creation
+      jobProps.remove(ConfigurationKeys.JOB_SCHEDULE_KEY);
+      jobProps.remove(ConfigurationKeys.JOB_NAME_KEY);
```
Comment (ibuenros) [inline gobblin-restli/gobblin-job-config-service/server/src/main/java/gobblin/service/JobConfigsResource.java]: I'm not sure there's a point to removing this. I think you're trying to make it so that doing a put followed by a get returns exactly what you put in, but why is this required?

File: gobblin-restli/gobblin-job-config-service/server/src/main/java/gobblin/service/JobConfigsResource.java [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+      jobProps.remove(ConfigurationKeys.JOB_GROUP_KEY);
+      jobProps.remove(ConfigurationKeys.JOB_TEMPLATE_PATH);
+      jobProps.remove(ConfigurationKeys.JOB_CONFIG_FILE_GENERAL_PATH_KEY);
```
Comment (ibuenros) [inline gobblin-restli/gobblin-job-config-service/server/src/main/java/gobblin/service/JobConfigsResource.java]: This should be removed by the catalog itself.

File: gobblin-restli/gobblin-job-config-service/server/src/main/java/gobblin/service/JobConfigsResourceFactory.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import com.linkedin.restli.server.resources.ResourceFactory;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang.Validate;
+
+
+/**
+ * JobConfigs Resource Factory that is used to create the JobConfigs restli resources.
+ */
+@Slf4j
+public class JobConfigsResourceFactory implements ResourceFactory {
```
Comment (ibuenros) [inline gobblin-restli/gobblin-job-config-service/server/src/main/java/gobblin/service/JobConfigsResourceFactory.java]: I would want this replaced for the `Guice` resource factory, we can discuss though.

File: gobblin-restli/server.gradle [hunk 11]
```diff
@@ -37,7 +37,9 @@ ext.apiProject = project(project.getParent().getPath() + ':api')
 
 dependencies {
   compile project(path: project.getParent().getPath() + ":api", configuration: 'dataTemplate')
+  compile project(path: project.getParent().getPath() + ":client")
```
Comment (ibuenros) [inline gobblin-restli/server.gradle]: why does server depend on client?

File: gobblin-restli/server.gradle [hunk 12]
```diff
@@ -37,7 +37,9 @@ ext.apiProject = project(project.getParent().getPath() + ':api')
...
   compile project(':gobblin-utility')
+  compile project(':gobblin-runtime')
```
Comment (ibuenros) [inline gobblin-restli/server.gradle]: This should not be a default dependency. Let's do the extra dependencies thing we talked about.

File: gobblin-restli/gobblin-job-config-service/client/src/main/java/gobblin/service/JobConfigClient.java [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+  private final JobconfigsRequestBuilders jobconfigsRequestBuilders;
+
+  /**
+   * Construct a {@link JobConfigClient} to communicate with http job config server at URI serverUri
+   * @param serverUri address and port of the REST server
+   */
+  public JobConfigClient(String serverUri) {
+    LOG.debug("JobConfigClient with serverUri " + serverUri);
+
+    httpClientFactory = Optional.of(new HttpClientFactory());
```
Comment (ibuenros) [inline gobblin-restli/gobblin-job-config-service/client/src/main/java/gobblin/service/JobConfigClient.java]: Either use underscores for instance variables, or use the keyword `this`.

File: gobblin-restli/gobblin-job-config-service/client/src/main/java/gobblin/service/JobConfigClient.java [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+  public void createJobConfig(JobConfig jobConfig)
+      throws RemoteInvocationException {
+    LOG.debug("createJobConfig with groupName " + jobConfig.getJobGroup() + " jobName " + jobConfig.getJobName());
+
+    CreateIdRequest<ComplexResourceKey<JobConfigId, EmptyRecord>, JobConfig> request =
+        jobconfigsRequestBuilders.create().input(jobConfig).build();
+    ResponseFuture<IdResponse<ComplexResourceKey<JobConfigId, EmptyRecord>>> jobConfigResponseFuture =
+        restClient.get().sendRequest(request);
+
+    jobConfigResponseFuture.getResponse();
```
Comment (ibuenros) [inline gobblin-restli/gobblin-job-config-service/client/src/main/java/gobblin/service/JobConfigClient.java]: Should you do something with the response code? What if the call failed?

**Comment by htran1** [comment]
This PR got messy, so moved the changes to #1589.
